core/src/epicli/cli/epicli.py

```diff
@@ -252,6 +258,10 @@ def upgrade_parser(subparsers):
                             help="Waits for all pods to be in the 'Ready' state before proceeding to the next step of the K8s upgrade.")
     optional.add_argument('--profile-ansible-tasks', dest='profile_ansible_tasks', action="store_true",
                             help='Enable Ansible profile_tasks plugin for timing tasks. (developer/debug option)')
+    optional.add_argument('--upgrade-components', dest='upgrade_components', type=str, required=False,
```

Comment:
Might I suggest to only have the ```--upgrade-components``` which should be default to something like ```"all"``` to specify to just upgrade all components? if its not all its a list of specific components you want to upgrade and there will be need for ```--upgrade-all```.

Finally check the ```choices``` argument for ```add_argument``` if you can add some validation for ```--upgrade-components``` argument.

Reply:
I agree with Luuk, probably this should work like this:
1. not pass --upgrade-components - this will work like upgrade all
2. --upgrade-components - this will fail if you won't specify components and upgrade only specified components

Reply:
it was the idea initially but changed. Wil amend it.

Reply:
> > > @ar3ndt @rpudlowski93 @erzetpe @seriva what do you think about making componet names case-insensitive for epicli? I would like to have no error if `Kafka` specified instead of `kafka`.
> > 
> > 
> > In my opinion It is definitely a good idea, I would like to have no errors in this scenario.
> 
> sure will do

IMO I don't mind either way. Kafka/kafka is the same for me. More important would be to add proper validation to the Epicli upgrade command what to accept and **validate** a , or space separated array. Argparse does not support that but a custom type would do the trick. Something allong the lines of:

```python
def upgrade_parser(subparsers):
    sub_parser = subparsers.add_parser('upgrade',
                                       description='Upgrades common and K8s components of an existing Epiphany Platform cluster.')
    optional = sub_parser._action_groups.pop()
    required = sub_parser.add_argument_group('required arguments')

    component_list = ['all', 'kubernetes', 'kafka']
    def comma_seperated_type(choices):
        """Return a function that splits and checks comma-separated values."""
        def splitarg(arg):
            values = arg.split(',')
            for value in values:
                if value not in choices:
                    raise argparse.ArgumentTypeError(
                        'invalid choice: {!r} (choose from {})'
                        .format(value, ', '.join(map(repr, choices))))
            return values
        return splitarg

    #required
    required.add_argument('-b', '--build', dest='build_directory', type=str, required=True,
                            help='Absolute path to directory with build artifacts.')

    #optional
    optional.add_argument('--offline-requirements', dest='offline_requirements', type=str, required=False,
                            help='Path to the folder with pre-prepared offline requirements.')
    optional.add_argument('--wait-for-pods', dest='wait_for_pods', action="store_true",
                            help="Waits for all pods to be in the 'Ready' state before proceeding to the next step of the K8s upgrade.")
    optional.add_argument('--profile-ansible-tasks', dest='profile_ansible_tasks', action="store_true",
                            help='Enable Ansible profile_tasks plugin for timing tasks. (developer/debug option)')
    optional.add_argument('--upgrade-components', dest='upgrade_components', type=comma_seperated_type(component_list), required=False,
                            help='Provides comma seperated list of components for upgrade selected from the following: [' + ','.join(map(str, component_list)) + ']')
    
    sub_parser._action_groups.append(optional)

    def run_upgrade(args):
        adjust_paths_from_build(args)
        with UpgradeEngine(args) as engine:
            return engine.upgrade()

    sub_parser.set_defaults(func=run_upgrade)
```

Also while you are making these changes maybe also straighten out the _- usage in the task file names:
![image](https://user-images.githubusercontent.com/11056665/109505070-859bd000-7a9c-11eb-8db0-4917a6627cac.png)